libraries/AP_HAL/CAN.h

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2017 Eugene Shamaev
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <inttypes.h>
+#include <string.h>
+#include <assert.h>
+#include <cmath>
+
+#include "AP_HAL_Namespace.h"
+#include "utility/functor.h"
+#include <uavcan/driver/can.hpp>
+#include <uavcan/time.hpp>
+
+class AP_UAVCAN;
+
+/**
+ * Single non-blocking CAN interface.
+ */
+class AP_HAL::CAN: public uavcan::ICanIface {
```

Comment:
The drivers are universal and can be used for any protocol. They are based on uavcan drivers because it is well defined and gives all functionality for any use case.